src/testrender/scene.cpp

```diff
@@ -0,0 +1,270 @@
+                          == shape.mesh.indices.size() / 3);
+        for (int i = 0, n = shape.mesh.indices.size(), f = 0; i < n;
+             i += 3, f++) {
+            int a = shape.mesh.indices[i + 0].position_index;
+            int b = shape.mesh.indices[i + 1].position_index;
+            int c = shape.mesh.indices[i + 2].position_index;
+            triangles.emplace_back(
+                TriangleIndices { base_idx + a, base_idx + b, base_idx + c });
+            int na = shape.mesh.indices[i + 0].normal_index;
+            int nb = shape.mesh.indices[i + 1].normal_index;
+            int nc = shape.mesh.indices[i + 2].normal_index;
+            // either all are valid, or none are valid
+            OSL_DASSERT((na < 0 && nb < 0 && nc < 0)
+                        || (na >= 0 && nb >= 0 && nc >= 0));
+            n_triangles.emplace_back(TriangleIndices {
+                na < 0 ? -1 : n_base_idx + na,
+                na < 0 ? -1 : n_base_idx + nb,
+                na < 0 ? -1 : n_base_idx + nc,
+            });
+            int ta = shape.mesh.indices[i + 0].texcoord_index;
+            int tb = shape.mesh.indices[i + 1].texcoord_index;
+            int tc = shape.mesh.indices[i + 2].texcoord_index;
+            OSL_DASSERT((ta < 0 && tb < 0 && tc < 0)
+                        || (ta >= 0 && tb >= 0 && tc >= 0));
+            uv_triangles.emplace_back(TriangleIndices {
+                ta < 0 ? -1 : uv_base_idx + ta, ta < 0 ? -1 : uv_base_idx + tb,
+                ta < 0 ? -1 : uv_base_idx + tc });
+
+            if (use_material_ids && !shape.mesh.material_ids.empty()) {
+                // remap the material ID to our indexing
+                int obj_mat_id = shape.mesh.material_ids[f];
+                OSL_ASSERT(obj_mat_id >= 0);
+                OSL_ASSERT(obj_mat_id < int(material_ids.size()));
+                shaderids.emplace_back(material_ids[obj_mat_id]);
+            } else
+                shaderids.emplace_back(shapeShaderID);
+            ntris++;
+        }
+        last_index.emplace_back(triangles.size());
+    }
+    double loadtime = timer();
+    errhandler.infofmt("Parsed {} vertices and {} triangles from {} in {}",
+                       nverts, ntris, filename,
+                       OIIO::Strutil::timeintervalformat(loadtime, 2));
+}
+
+void
+Scene::add_sphere(const Vec3& c, float r, int shaderID, int resolution)
+{
+    const int W    = 2 * resolution;
+    const int H    = resolution;
+    const int NV   = 2 + W * H;  // poles + grid = total vertex count
+    int base_idx   = verts.size();
+    int n_base_idx = normals.size();
+    int t_base_idx = uvs.size();
+    // vertices
+    verts.emplace_back(c + Vec3(0, r, 0));  // pole +z
+    normals.emplace_back(0, 1, 0);
+    // W * H grid of points
+    for (int y = 0; y < H; y++) {
+        float t = float(y + 0.5f) / float(H);
+        float z = cosf(t * float(M_PI));
+        float q = sqrtf(1 - z * z);
```

Comment:
`float q = sqrtf(1 - z * z);`
So if z > 1.0 by even 1ulp, then sqrtf could result in a NAN.
The result of cosf could allow a certain # of bits accuracy which means it could have results slightly greater than 1.0f.
Please try clamping the input to 0.f
`float q = sqrtf(std::max(0.f, 1 - z * z));`

and see if you NAN issues go away.

Reply:
Sometimes in C++ I really miss OSL's "debug_nan" feature that will insert extra code around all math to 100% guarantee identifying the line in which a NaN first creeps in.

Reply:
I was just did source code review to find that, C library does not guarantee perfect accuracy, so some clamping is needed if feeding values into a NaN producing operation.

Would someone be willing to try that and re-enable icpx CI?

Reply:
Yeah, I will try right now!

This is a lesson we know well, that's why OIIO::safe_sqrt() exists!

Reply:
running now...

Reply:
I believe this is a smart change regardless, so I will submit the PR.

But I'm afraid it doesn't solve the problem. A number of tests still either time out (800 seconds!) or hit assertions. Only for icx.